# Understanding the Dataset

In [262]:
import pandas as pd

In [ ]:
# Importing the dataset

df = pd.read_csv("../data/Titanic.csv")
df1=df.copy() # Making copy for further operations

### 1. What is the size of the data ?

In [283]:
print(f"Shape of the data : {df.shape}")

Shape of the data : (891, 12)


### 2. How does the data looks like ?


In [284]:
# Gives first 5 data sets  
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [285]:
# Gives randomly selected number of data 
df.sample(5) # top 5 rows selected randomly to check the nature of data 

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
726,727,1,2,"Renouf, Mrs. Peter Henry (Lillian Jefferys)",female,30.0,3,0,31027,21.0000,NaN,S
612,613,1,3,"Murphy, Miss. Margaret Jane",female,NaN,1,0,367230,15.5000,NaN,Q
620,621,0,3,"Yasbeck, Mr. Antoni",male,27.0,1,0,2659,14.4542,NaN,C
253,254,0,3,"Lobb, Mr. William Arthur",male,30.0,1,0,A/5. 3336,16.1000,NaN,S
176,177,0,3,"Lefebre, Master. Henry Forbes",male,NaN,3,1,4133,25.4667,NaN,S


### 3. What is the data-type of columns ?


In [286]:
# Gives the whole picture of data type used and null-count helps to track the missing values 
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


### 4. Are there any missing values ?


In [287]:
# It gives the boolean data whether the data is missing or not!
df.isnull()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,True,False
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False


In [288]:
# using .sum() we get number of that missing values in each columns
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 5. How does the data look mathematically ?


In [289]:
# Provide the mathematical intuition of the data like mean, std, min, max etc..
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### 6. Are there any duplicate values ?


In [290]:
# Gives boolean values for each rows indicating duplicate values
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

In [291]:
# using .sum() given the values of duplicated values 
df.duplicated().sum()

np.int64(0)

### 7. How many unique values do the high-cardinality columns have?

In [292]:
print(f"Number of unique Tickets : {df['Ticket'].nunique()}")
print(f"Number of Unique Names : {df['Name'].nunique()}")
print(f"Number of unique Cabin : {df['Cabin'].nunique()}")

# short way : df[['Name','Ticket','Cabin']].nunique()

Number of unique Tickets : 681
Number of Unique Names : 891
Number of unique Cabin : 147


### 8. Are there correlation b/w columns ?

In [293]:
# As in this data sets Sex is the single strongest predictor of survival in this dataset ("women and children first" pattern)
df['Sex'] = df['Sex'].map({'male':0,'female':1})
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",0,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",1,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",1,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",0,26.0,0,0,111369,30.0000,C148,C


In [294]:
# Shows the correlation b/w each columns with respect to each other
# NOTE : Run this column at last 
df = df.drop(columns=['Name','Ticket','Cabin', 'Embarked','PassengerId'])


In [295]:
df.corr()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare
Survived,1.000000,-0.338481,0.543351,-0.077221,-0.035322,0.081629,0.257307
Pclass,-0.338481,1.000000,-0.131900,-0.369226,0.083081,0.018443,-0.549500
Sex,0.543351,-0.131900,1.000000,-0.093254,0.114631,0.245489,0.182333
Age,-0.077221,-0.369226,-0.093254,1.000000,-0.308247,-0.189119,0.096067
SibSp,-0.035322,0.083081,0.114631,-0.308247,1.000000,0.414838,0.159651
Parch,0.081629,0.018443,0.245489,-0.189119,0.414838,1.000000,0.216225
Fare,0.257307,-0.549500,0.182333,0.096067,0.159651,0.216225,1.000000


### 9. Extracting title from Name

In [299]:
# Extracting the Title of the name to see verity of persons 

df1['Title'] = df1['Name'].str.extract(r',\s*([^.]*)\.')
df1['Title'].value_counts()

Title
Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Major             2
Mlle              2
Col               2
Don               1
Mme               1
Ms                1
Lady              1
Sir               1
Capt              1
the Countess      1
Jonkheer          1
Name: count, dtype: int64

### 10. Survival rate by category 
— Pearson correlation isn't the right tool for categorical columns like Sex, Pclass, Embarked. Use groupby instead.

In [300]:
# Correlation survival with respect to sex
# Female survival is larger as priority was "women and children first" 
df1.groupby('Sex')['Survived'].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

In [301]:
# Correlation of Pclass with respect to Survival 
df1.groupby('Pclass')['Survived'].mean()

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

In [302]:
# Correlation of Embarked with respect to Survival
df1.groupby('Embarked')['Survived'].mean()

Embarked
C    0.553571
Q    0.389610
S    0.336957
Name: Survived, dtype: float64

Sex turns out to be the single strongest signal in this dataset (74% vs 19% survival) — it got missed in the original notebook because it was dropped *before* checking correlation.

### 11. Fare — checking skew/outliers, not just mean/std

In [303]:
df1['Fare'].skew()

np.float64(4.787316519674893)

A skew of ~4.8 is heavily right-skewed (0 = symmetric). Combined with std (49.7) being larger than the mean (32.2) from `describe()`, this confirms outliers — there's a fare as high as 512 against a median of just 14.45. Worth a log-transform or capping before modeling.

### 12. Cabin is 77% missing — instead of dropping it outright, turn it into a signal

In [305]:
df1['Has_Cabin'] = df1['Cabin'].notnull().astype(int)
df1.groupby('Has_Cabin')['Survived'].mean()

Has_Cabin
0    0.299854
1    0.666667
Name: Survived, dtype: float64

Having a recorded cabin is itself predictive (67% vs 30% survival) — likely a proxy for class/fare. This turns a mostly-missing column into a usable binary feature instead of throwing it away.